Swiften/MIX/MIXRegistry.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) 2017 Tarun Gupta
+ * Licensed under the simplified BSD license.
+ * See Documentation/Licenses/BSD-simplified.txt for more information.
+ */
+
+#include <Swiften/MIX/MIXRegistry.h>
+
+namespace Swift {
+
+MIXRegistry::MIXRegistry(const JID& ownJID, IQRouter* iqRouter, XMPPRoster* xmppRoster) : ownJID_(ownJID), iqRouter_(iqRouter), xmppRoster_(xmppRoster), rosterPopulated_(false) {
+    xmppRoster_->onInitialRosterPopulated.connect(boost::bind(&MIXRegistry::handleRosterPopulated, this));
+    xmppRoster_->onJIDAdded.connect(boost::bind(&MIXRegistry::handleJIDAdded, this, _1));
+    xmppRoster_->onJIDRemoved.connect(boost::bind(&MIXRegistry::handleJIDRemoved, this, _1));
+}
+
+MIXRegistry::~MIXRegistry() {
+
+}
+
+void MIXRegistry::joinChannel(const JID& channelJID, const std::unordered_set<std::string>& nodes) {
+    auto joinPayload = std::make_shared<MIXJoin>();
+    joinPayload->setChannel(channelJID);
+    for (auto node : nodes) {
+        joinPayload->addSubscription(node);
+    }
+    auto request = std::make_shared<GenericRequest<MIXJoin>>(IQ::Set, ownJID_.toBare(), joinPayload, iqRouter_);
+    request->onResponse.connect(boost::bind(&MIXRegistry::handleJoinResponse, this, _1, _2));
+    request->send();
+}
+
+void MIXRegistry::handleJoinResponse(MIXJoin::ref /*payload*/, ErrorPayload::ref error) {
+    if (error) {
+        onChannelJoinFailed(error);
+    }
+}
+
+void MIXRegistry::leaveChannel(const JID& channelJID) {
+    auto leavePayload = std::make_shared<MIXLeave>();
+    leavePayload->setChannel(channelJID);
+    auto request = std::make_shared<GenericRequest<MIXLeave>>(IQ::Set, ownJID_.toBare(), leavePayload, iqRouter_);
+    request->send();
+}
+
+void MIXRegistry::handleRosterPopulated() {
+    rosterPopulated_ = true;
+}
+
+void MIXRegistry::handleJIDAdded(const JID& jid) {
+    if (rosterPopulated_ && xmppRoster_->isChannel(jid)) {
+        std::map<JID, MIX::ref>::iterator i = entries_.find(jid);
+        if (i == entries_.end()) {
+            entries_.insert(std::make_pair(jid, std::make_shared<MIXImpl>(ownJID_, jid, iqRouter_)));
+        }
+        onChannelJoined(jid);
+    }
+}
+
+void MIXRegistry::handleJIDRemoved(const JID& jid) {
+    if (rosterPopulated_ && xmppRoster_->isChannel(jid)) {
+        std::map<JID, MIX::ref>::iterator i = entries_.find(jid);
+        if (i != entries_.end()) {
+            entries_.erase(jid);
+        }
+        onChannelLeaved(jid);
+    }
+}
+
+std::unordered_set<std::string> MIXRegistry::getJoinedChannels() {
```

Comment:
I'd have expected that, if you request the mix bits of the roster, it'd then sort it all out for you populate them and you'd then use them. Is that not the likely use case?